File: leetcode/39_Combination Sum.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        # 중복 조합. 합이 정해짐
        # 백트래킹 - 정렬 후 넘어서면
        candidates.sort()
        ans=[]
        def dfs(start,sumv,stack,ans):
            if sumv > target:
                return 
            
            if sumv==target:
                ans+=[tuple(stack)]
            
            
            for i in range(start,len(candidates)):
                stack+=[candidates[i]]                
                dfs(i,sumv+candidates[i],stack,ans)
                stack.pop()
                
        dfs(0,0,[],ans)
        return ans

File: leetcode/39_Combination Sum.cpp

class Solution {
public:
    vector<vector<int>> ans; 
    vector<int> partial; 

    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        backTracking(candidates, 0, target);
        return ans;
    }

    void backTracking(vector<int>& candidates, int currIdx, int target) {
        if (target==0) {
            ans.push_back(partial);
            return; 
        }

        if (target<0 || currIdx==candidates.size()) {// pruning 
            return; 

        partial.push_back(candidates[currIdx]);
        backTracking(candidates, currIdx, target-candidates[currIdx]);
        partial.pop_back();
        backTracking(candidates, ++currIdx, target);
    }
};